server/src/test/java/org/elasticsearch/index/query/MatchPhraseQueryBuilderTests.java

```diff
@@ -192,36 +192,40 @@ public void testParseFailsWithMultipleFields() throws IOException {
 
     public void testRewriteToTermQueries() throws IOException {
         QueryBuilder queryBuilder = new MatchPhraseQueryBuilder(KEYWORD_FIELD_NAME, "value");
-        SearchExecutionContext context = createSearchExecutionContext();
-        QueryBuilder rewritten = queryBuilder.rewrite(context);
-        assertThat(rewritten, instanceOf(TermQueryBuilder.class));
-        TermQueryBuilder tqb = (TermQueryBuilder) rewritten;
-        assertEquals(KEYWORD_FIELD_NAME, tqb.fieldName);
-        assertEquals(new BytesRef("value"), tqb.value);
+        for (QueryRewriteContext context : new QueryRewriteContext[] { createSearchExecutionContext(), createQueryRewriteContext() }) {
```

Comment:
If I understand correctly here we test if the rewrite happens no matter when the actual rewrite is called. This is to make sure that we do the rewrite as soon as possible and in the worst case scenario we do it as before, rewriting with a `SearchExecutionContext` including an `IndexSearcher`. In our K8s tests this rewrite will happen before anyway due to the new logic, but if for some reason (for instance condition on shard pre-filtering is not met) then we will at least rewrite it later.

Reply:
Right, this is to ensure that both rewrite with search execution context and just a rewrite context are able to rewrite these queries correctly. The k8s o11y case we can skip a lot of shards, because the rewrite query context rewrite was already able to rewrite to a match no docs query.